mm2src/mm2_main/src/lp_swap/maker_swap.rs

```diff
@@ -1196,91 +1195,88 @@ impl MakerSwap {
-                    MakerSwapEvent::MakerPaymentRefundFailed(
-                        ERRL!("!maker_coin.wait_for_htlc_refund: {}", err.to_string()).into(),
-                    ),
-                ])),
-            };
-        }
-
+        // Keep trying to recover funds (by refunding the maker payment or spending the taker payment) until successful or face an irrecoverable error.
         loop {
-            match self.maker_coin.can_refund_htlc(locktime).await {
-                Ok(CanRefundHtlc::CanRefundNow) => break,
-                Ok(CanRefundHtlc::HaveToWait(to_sleep)) => Timer::sleep(to_sleep as f64).await,
-                Err(e) => {
-                    error!("Error {} on can_refund_htlc, retrying in 30 seconds", e);
-                    Timer::sleep(30.).await;
+            match self.recover_funds().await {
+                // We recovered the swap successfully.
+                Ok(recovered_swap) => match recovered_swap {
+                    // We recovered the swap by refunding the maker payment.
+                    RecoveredSwap {
+                        action: RecoveredSwapAction::RefundedMyPayment,
+                        coin: _,
+                        transaction,
+                    } => {
+                        let tx_ident = TransactionIdentifier {
+                            tx_hex: transaction.tx_hex().into(),
+                            tx_hash: transaction.tx_hash_as_bytes(),
+                        };
+                        info!("Maker payment refund tx {:02x}", tx_ident.tx_hash);
+                        return Ok((Some(MakerSwapCommand::FinalizeMakerPaymentRefund), vec![
+                            MakerSwapEvent::MakerPaymentRefunded(Some(tx_ident)),
+                        ]));
+                    },
+                    // We recovered the swap by proceeding forward and spending the taker payment. The swap wasn't actually a failure.
+                    // Roll back to confirming the taker payment spend.
+                    RecoveredSwap {
+                        action: RecoveredSwapAction::SpentOtherPayment,
+                        coin: _,
+                        transaction,
+                    } => {
+                        let tx_ident = TransactionIdentifier {
+                            tx_hex: transaction.tx_hex().into(),
+                            tx_hash: transaction.tx_hash_as_bytes(),
+                        };
+                        info!("Refund canceled. Taker payment spend tx {:02x}", tx_ident.tx_hash);
+                        // TODO: We prepared for refund but didn't finalize refund. This must be breaking something for lightning.
+                        return Ok((Some(MakerSwapCommand::ConfirmTakerPaymentSpend), vec![
+                            MakerSwapEvent::TakerPaymentSpent(tx_ident),
+                            MakerSwapEvent::TakerPaymentSpendConfirmStarted,
+                        ]));
+                    },
```

Comment:
Can be simplified using closure and `return match`

```rust
    // We recovered the swap successfully.
                Ok(recovered_swap) => {
                    let tx_ident = |tx: TransactionEnum| TransactionIdentifier {
                        tx_hex: tx.tx_hex().into(),
                        tx_hash: tx.tx_hash_as_bytes(),
                    };

                    // We recovered the swap by refunding the maker payment.
                    return match recovered_swap.action {
                        RecoveredSwapAction::RefundedMyPayment => {
                            let ident = tx_ident(recovered_swap.transaction);
                            info!("Maker payment refund tx {:02x}", ident.tx_hash);
                            Ok((Some(MakerSwapCommand::FinalizeMakerPaymentRefund), vec![
                                MakerSwapEvent::MakerPaymentRefunded(Some(ident)),
                            ]))
                        },
                        // We recovered the swap by proceeding forward and spending the taker payment. The swap wasn't actually a failure.
                        // Roll back to confirming the taker payment spend.
                        RecoveredSwapAction::SpentOtherPayment => {
                            let ident = tx_ident(recovered_swap.transaction);
                            info!("Refund canceled. Taker payment spend tx {:02x}", ident.tx_hash);
                            // TODO: We prepared for refund but didn't finalize refund. This must be breaking something for lightning.
                            Ok((Some(MakerSwapCommand::ConfirmTakerPaymentSpend), vec![
                                MakerSwapEvent::TakerPaymentSpent(ident),
                                MakerSwapEvent::TakerPaymentSpendConfirmStarted,
                            ]))
                        },
                    };
                },
```

Reply:
[5e0709d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2280/commits/5e0709d22a683c3304019872885aeaf438feb20f)